### ISSUE TITLE
Virtual Machine hotplug disk support

### ISSUE THREAD

**Issue by futuretea** [harvester/harvester#283]
Currently, `Attached VM`  filed display on the volume page does not mean it is mounted by the VM

KVM can hotplug disk, in the latest kubevirt release `v0.36.0-rc.0`, kubevirt can hotplug DataVolume and PVC volumes into a running Virtual Machine

links:
kubevirt/kubevirt#4250
kubevirt/kubevirt#4458

---

### Development Tasks

- Backend API 
  - [x] New VM action `addVolume`
    - #990
  - [x] New VM action `removeVolume`
    - #990
- Web UI
  - [x] New menu item and popup form for `addVolume` action
  - [ ] support volume hot-unplug
  - [ ] <strike>Differentiate hotplug volumes from others in VM's volume section</strike>
  - [ ] <strike>❗ **Discusssion needed:** Override remove button in VM's volume section with a confirm popup for `removeHotPlugDisk` action</strike>
- Backup/restore support (by leveraging Harvester's existing backup/restore feature)
  - [x] ❌ ~~For persistent hotplug volumes~~ 
    - Future version of KubeVirt would support hot-unplugg volumes after VM restarted, so no need to modify any logic
  - [x] ❌ ~~For ephemeral hotplug volumes~~
    - Wouldn't support this since we haven't found a good use case for this feature. And it will add too much complexity
    - Future version of KubeVirt would support hot-unplugg volumes after VM restarted, so no need to modify any logic
  - [x] 🐛 #987
    - #997 
  - [x] 🐛 #985
    - #986
- Upgrade to latest KubeVirt
  - [x] Upgrade to KubeVirt 0.42.1 or later: Since there is a bug introduced in libvirt 6.5.0-2([^1] [^2] [^3]), and it seems that 6.6.0 does not fix it. Unfortunately, harvester [uses KubeVirt 0.40.0] which is based on [libvirt 6.6.0].
    - #1068
  - [x] Upgrade to new KubeVirt version that supports live migrations and hotplug auto-reattachment
- Live migration support
  - [x] 🛑 **Blocker:** kubevirt/kubevirt#5728
  - [x] 🛑 **Blocker:** rancher/rke2#1204
  

### Ephemeral v.s. Persistent Hotplug Volume

So here I need to clarify what's the difference between "ephemeral" and "persistent" hotplug volume. For KubeVirt, [VM and VMI API handles `AddVolume` request][add-vol-req] separately. If one adds a hotplug volume against VMI, we call it "ephemeral" because the volume won't re-attach after VM restart. Ephemeral hotplug volume is not snapshottabble for Harvester backup logic since the current logic backups content inside VM but not VMI.

For persistent volume, if a `AddVolume` request is sent to VM resource, KubeVirt modifies the VM spec by adding hotplug volume as a manually attached volume. Therefore, any future recreate/restart/migrate will reflect the changes. In addition, although we send the request against VM resource, the VMI's status still the same as if we send the request against VMI directly. Note that once the VM goes restarted, the change becomes permanent and one cannot hot-unplugged the volume anymore.

[add-vol-req]: https://github.com/kubevirt/kubevirt/blob/3e4ea2d9b8d4405dc307e885bc49f10c8ce5372c/pkg/virt-api/rest/subresource.go#L1108-L1126
[libvirt 6.6.0]: https://github.com/kubevirt/kubevirt/blob/v0.40.0/hack/rpm-deps.sh#L8
[uses KubeVirt 0.40.0]: https://github.com/harvester/harvester/blob/84781dcbf13337c9b0711551910ecc5789e4ee93/deploy/charts/harvester/values.yaml#L35
[^1]: https://bugs.archlinux.org/task/67921#comment193002
[^2]: https://www.reddit.com/r/archlinux/comments/iumubw
[^3]: https://www.reddit.com/r/qemu_kvm/comments/ittaez

**Comment by bk201** [comment]
I can confirm the feature is available in KubeVirt `0.38.1` that is currently run with Harvester.

Here are some experiments:
- Add this ConfigMap to the `harvester-system` namespace, this enables hot plugging support.
  ```yaml
  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: kubevirt-config
    namespace: harvester-system
    labels:
      kubevirt.io: ""
  data:
    feature-gates: "HotplugVolumes" 
   ```
- Create a volume on the `Volumes` page
- Attach the volume via [API call](https://kubevirt.io/api-reference/v0.38.1/operations.html#_v1vm-addvolume) or my [WIP change](https://github.com/bk201/kubevirt/commits/virtctl-hotplugging) to virtctl.
- The new-created volume should be available in the instance.
- Detaching volume works too.

## Some observations:
- The plugging state is reflected in the Harvester GUI
![Screen Shot 2021-03-17 at 2 45 19 PM](https://user-images.githubusercontent.com/1691518/111426387-90f32a80-872f-11eb-98b7-b2a193de008e.png)

- The hot-plugged volume can be seen as `Existing Volume` in the `Virtual Machine` detail -> `Volumes` tab
![Screen Shot 2021-03-17 at 2 45 35 PM](https://user-images.githubusercontent.com/1691518/111426383-8df83a00-872f-11eb-90a2-821b438fdad2.png)

- I shut down the cluster and restart the cluster again, the hotplugging mechanism doesn't seem to work. The attached volumes are not restored. Need more investigation on this.


## Limitation
From the [design document](https://docs.google.com/document/d/14WitxXoScSwI459DTm4PVRayDHL3CbmlB2PKSPQ66WY/edit?usp=sharing) and some talks, there are some limitations.
- Live migration of a hot-plugged VM is not supported yet.
- Online snapshot is not supported.

**Comment by guangbochen** [comment]
I think we should enable the `hotplug` feature by default and configure the following feature gate at kubevirt CR: https://github.com/rancher/harvester/blob/master/deploy/charts/harvester/dependency_charts/kubevirt/values.yaml#L127
```
featureGates: ["LiveMigration", "HostDisk"]
```

**Comment by yasker** [comment]
@bk201 @guangbochen Since live migration is not supported, we probably cannot use the hotplug disk feature from KubeVirt directly.

**Comment by yasker** [comment]
To clarify the experience, we want to achieve something like this:

1. User creates a VM
2. User attaches a disk to the VM when the VM is running, e.g. data storage.
3. Due to the node maintenance, the VM needs to be migrated to other nodes.
4. VM should be migrated to another node correctly.
5. User can choose to detach the volume or just shutdown the VM.
    1. If the user decides to start the VM again without detaching the volume, the volume should be attached automatically after.
6. User will be allowed to detach any volumes except for the root volume when the VM is running.

Also, the current hotplug design used one pod per volume, which might result in one physical node having too many volumes due to it's hosting multiple VMs. It's better if we can solve that problem too.

**Comment by innobead** [comment]
ref: https://github.com/kubevirt/kubevirt/pull/5728

**Comment by weihanglo** [comment]
Also One pod for all hotplug volumes: https://github.com/kubevirt/kubevirt/pull/5649

**Comment by weihanglo** [comment]
According to [this thread][1], live migration with block volumes may fail in a K8s cluster with CPU Manager and static policy. There is a fix https://github.com/opencontainers/runc/commit/bf7492ee5 but in the thread they presume the fix may be released in K8s 1.22.

[1]: https://groups.google.com/g/kubevirt-dev/c/0UbMD2R92mQ

**Comment by yasker** [comment]
Filed https://github.com/rancher/rke2/issues/1204 to track the file in RKE2.

**Comment by weihanglo** [comment]
Jotted down some design questions:


1. KubeVirt's virtctl provides a persistent feature for hotplug volume, which means [whether write the changes to VM template or not][1]. If choosing persisnt, hotplug volumes will "become a persistent volume” after a VM restart. Should we give the options to user? Or just choose one default way.

2. Since Harvester uses in-house snapshot instead of KubeVirt snapshot, we can support that for hotplug volume. The questions is, should hotplug volumes be able to be backuped and restored as other normal volumes in Harvester?

The second question depends on the first because `non-persistent` hotplug volume will only write changes to VM instance, and VMI changes are not tracked by current backup/restore logic.

[1]: https://github.com/kubevirt/kubevirt/blob/e3c8403c7913d171580d1f7c70934f9f04e850dc/pkg/virtctl/vm/vm.go#L297-L301

---

Besides, here is some mock UI. The expected frontend flow to add a hotplug volume may be

1. Click "Add Hotplug Volume" button
2. Popup show
3. Fill out the form and click "Create"
4. Frontend sends an HTTP request to `v1/kubevirt.io.virtualmachines/<ns>/vm-3?action=addhotplugdisk` with a new `addhotplugdisk` action

In addtion, for hotplug volume removal, it needs a confirm dialog as well.

<img width="939" alt="image" src="https://user-images.githubusercontent.com/14314532/123925901-f1943380-d9bd-11eb-8099-bbb2857160d1.png">

<img width="915" alt="image" src="https://user-images.githubusercontent.com/14314532/123925864-e6410800-d9bd-11eb-9226-85875ff87b03.png">

**Comment by guangbochen** [comment]
1. the type should default to persisted when adding a new hot plugin disk, and if we want to support the ephemeral disk, can we tell users the disk type from the UI or they can only check within the VM?
2. we should support both snapshot and backup of the persisted disk since this is just a regular disk of the Harvester VM.

In regards to the UI, I think we only need to add an additional VM action.  Because editing the VM will trigger the VM to reboot and this is not what hotplug expected to.

**Comment by weihanglo** [comment]
Thanks for your rapid reply. Totally agree with the UI part, which makes the intention more clear. 👍🏾

If we want to show "Persistent" or "Ephemeral" tag for hotplug volume, frontend need to 

1. first inspect the spec and status of virtualmachine to determine if a volume has datavolumetemplate, and
2. then inspect the status of virtualmachineinstance to determine if a volume is hotplug or not

I'll update more details if this is in the plan.

For the snapshot/backup stuff. Did you mean for ephemeral hotplug disk there is no need to do snapshot backup even the disk is attaching at the moment?

**Comment by yasker** [comment]
@guangbochen @weihanglo I doubt we need use to choose `persistent` or `ephemeral`. What is the use case for ephemeral? 

FYI: not directly related, but we're removing the snapshot feature from GA scope due to complexity on Harvester and Longhorn side.

**Comment by weihanglo** [comment]
> @guangbochen @weihanglo I doubt we need use to choose `persistent` or `ephemeral`. What is the use case for ephemeral? 

Yeah. We also tend to skip `ephemeral` at this moment.

> but we're removing the snapshot feature from GA scope

What do you refer to by snapshot here? When I say snapshot I mean Harvester backup/restore feature. Are we on the same page? Sorry for misusing these terms.

**Comment by yasker** [comment]
@weihanglo I meant https://github.com/harvester/harvester/issues/553 . It's using Longhorn in-cluster snapshot (instead of backup).

**Comment by gitlawr** [comment]
There seems to be gaps between kubevirt behaviors and our expectations:

~~**Kubevirt behavior:**
If the volume is added ephemerally(in VMI spec), after a VM restart the disk is detached.
**Expected:**~~
>If the user decides to start the VM again without detaching the volume, the volume should be attached automatically after.

Edit: We don't consider the ephemeral case

**Kubevirt behavior:**
The number of permanent volumes cannot be changed without a restart.
If the volume is added as permanent volumes(in VM spec), after a VM restart the disk cannot be detached.
**Expected:**
>User will be allowed to detach any volumes except for the root volume when the VM is running.

**Comment by weihanglo** [comment]
Related "Support hotplug volume being hot-unpluggable after restart": https://github.com/kubevirt/kubevirt/issues/6010

**Comment by weihanglo** [comment]
Note that some of our current logic depends on CDI client v1alpha1 while using KubeVirt `vm.spec.dataVolumeTemplate`, but Harvester plan to remove CDI dependency (#1058). Therefore, the upgrade of KubeVirt would be blocked by #1058.

**Comment by yasker** [comment]
@weihanglo What's the update on this issue? Does this mean the guest CSI driver won't actually work since we cannot hot unplug the volume?

**Comment by weihanglo** [comment]
Ｗe're currently upgrading KubeVirt to v0.4.5 (#1336) and only after that we can hot unplug volume. AFAIK the hot unplug API in backend side is completed but the UI side seems lack the feature.

**Comment by yasker** [comment]
UI part is fine. Let's file another issue for UI to hot unplug the volume. It needs to work for the CSI driver from the backend.

**Comment by guangbochen** [comment]
verified in `master-82506f55-head` and volume hot-plugin feature is supported, volume hot-unplug feature will be tracked on the https://github.com/harvester/harvester/issues/1401.
